src/main/java/core/basesyntax/Car.java

```diff
@@ -1,13 +1,87 @@
 package core.basesyntax;
 
+import java.util.ArrayList;
 import java.util.List;
+import java.util.Objects;
 
 /**
  * Make this class immutable. See requirements in the README file
  */
-public class Car {
-    private int year;
-    private String color;
-    private List<Wheel> wheels;
-    private Engine engine;
+
+public final class Car {
+    private final int year;
+    private final String color;
+    private final List<Wheel> wheels;
+    private final Engine engine;
+
+    public Car(int year, String color, List<Wheel> wheels, Engine engine) {
+        this.year = year;
+        this.color = color;
+        this.wheels = wheelsClone(wheels);
+        this.engine = engineClone(engine);
+    }
+
+    public int getYear() {
+        return year;
+    }
+
+    public String getColor() {
+        return color;
+    }
+
+    public List<Wheel> getWheels() {
+        return wheelsClone(wheels);
+    }
+
+    public Engine getEngine() {
+        return engineClone(engine);
+    }
+
+    public Car changeEngine(Engine engine) {
+        return new Car(year, color, wheels, engine);
+    }
+
+    public Car changeColor(String newColor) {
+        return new Car(year, newColor, wheels, engine);
+
+    }
+
+    public Car addWheel(Wheel newWheel) {
+        List<Wheel> wheelsList = getWheels();
+        wheelsList.add(newWheel);
+        return new Car(year, color, wheelsList, engine);
+
```

Comment:
```suggestion
    public Car addWheel(Wheel newWheel) {
        List<Wheel> wheelsList = getWheels();
        wheelsList.add(newWheel);
        return new Car(year, color, wheelsList, engine);
```
you are making deep twice here, in the constructor and in getWheels(); let's first create a new car and then add a wheel clone to its wheels

Reply:
done